src/parser/lexer.rs

```diff
@@ -0,0 +1,140 @@
+use std::{iter::Peekable, str::Chars};
+
+use super::token::Token;
+use super::range::{Range, Position};
+
+/// The lexer state
+#[derive(Debug)]
+pub struct Lexer<'a> {
+    pub input: &'a str,
+    pub position: usize,
+    pub peekable: &'a mut Peekable<Chars<'a>>
+}
+
+fn is_whitespace(c: char) -> bool {
+    matches!(c, ' ' | '\r' | '\t')
+}
+
+impl<'a> Lexer<'a> {
+    pub fn new(
+        input: &'a str,
+        peekable: &'a mut Peekable<Chars<'a>>
+    ) -> Lexer {
+        Lexer {
+            input,
+            position: 0,
+            peekable,
+        }
+    }
+
+    pub fn tokenize(&mut self) -> (Token, Range) {
+        let start = self.position;
+
+        match self.peekable.peek() {
+            None => (Token::EOF, self.mk_range(start)),
+            Some(c) => match c {
+                c if is_whitespace(*c) => {
+                    self.accumulate_while(&is_whitespace);
+                    self.tokenize()
+                }
+                // TODO: Add other types of numbers on it too
+                c if c.is_ascii_digit() => {
+                    let num = self.accumulate_while(&|x| x.is_ascii_digit() || x == '_');
+                    (Token::LInt(num.replace('_', "").parse().unwrap()), self.mk_range(start))
```

Comment:
`1____________________0` = 10? :P

Reply:
Also, that `unwrap` may trigger an error that instead should be handled accordingly.

Reply:
I'm not against the "1_______0". It's something that is highly unlikely to happen in a serious code and the way it is being parsed rn looks like the simplest way to make something like "1_000_000" work. UUr but It seems that it might trigger an error in case of a big enough number lol so we have to handle it.